doc/man/man4/mstr.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/mstr.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH MSTR 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B mstr
\- \fISTREAMS\fR multi-stream module
.\"
.\"
.SH SYNOPSIS
.PP
.HP 8
.BI "int " fd ;
.PD 0
.HP
.BI "int open(" \*(lq\fR/dev/nuls\fP\*(rq ", " \fR0\fP );
.HP
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqmstr\*(rq );
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B mstr
is a
.I STREAMS
module that is pushed over any
.IR Stream ,
.IR STREAMS -based
.BR pipe (4)
or
.IR STREAMS -based
.BR fifo (4s).
.PP
Aside from being a
.I STREAMS
module,
.B mstr
performs no actions other than to convert
.BR M_PROTO "(9) and " M_PCPROTO (9)
messages received from above into
.BR M_CTL "(9) and " M_PCCTL (9)
messages pass downward (resp.); and visa versa in the upstream direction.
That is, each
.BR M_PROTO (9)
message received from above is changed to a
.BR M_CTL (9)
message type and passed downstream without queuing nor consideration for flow
control.  Also, any
.BR M_CTL (9)
message received from below is converted to a
.BR M_PROTO (9)
message and passed upstream.  The same is also true for
.BR M_PCPROTO (9)
and
.BR M_PCCTL (9)
messages.
.\"
.\"
.SH USAGE
.PP
The primary purpose of the
.B mstr
module is to make the internal multi-stream interface provided by some
standard drivers available to user-space programs.  The module is simply
pushed over any driver that implements the multi-stream internal
.BR M_CTL "(9) or " M_PCCTL (9)
interface, or that otherwise requires conversion to and from
.BR M_PROTO "(9) or " M_PCPROTO (9)
and
.BR M_CTL "(9) or " M_PCCTL (9)
messages.
.\"
.\"
.SH NOTICES
.PP
.BR mstr
is implementation-specific to the
.B @PACKAGE_TITLE@
package.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH MODULES
.PP
.B mstr
is a
.I STREAMS
module named
.IR \*(lqmstr\*(rq .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR open (2s),
.BR ioctl (2s),
.BR streamio (7),
.BR putnext (9),
.BR M_PROTO (9),
.BR M_PCPROTO (9),
.BR M_CTL (9),
.BR M_PCCTL (9).
.\"
.\"
.SH BUGS
.PP
.B mstr
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B mstr
is compatible with the modules implementing the implementation-specific
multi-stream interface, or in any case where it is necessary to convert
.BR M_PROTO "(9) and " M_PCPROTO (9)
messages to and from
.BR M_CTL "(9) and " M_PCCTL (9)
across the module.
.IP \(em \w'\(em\(em'u
.B mstr
is specific to
.BR @PACKAGE_TITLE@ ,
and will not be used by portable
.I STREAMS
modules, drivers and applications programs.
.IP \(em
The name
.B mstr
is not standard and was chosen for
.BR @PACKAGE_TITLE@ .
.IP \(em
.B mstr
is used primarily by the
.BR test-rtp (8)
conformance test program.  It will not be used by portable
.I STREAMS
modules, drivers and applications.
.PP
See
.BR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.B mstr
is implementation-specific and is not subject to standardization.
.\"
.\"
.SH HISTORY
.PP
.B mstr
first appeared in
.B @PACKAGE_TITLE@
in release
.BR openss7-1.1.1-20101117 .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/mstr.c
/*****************************************************************************

 @(#) File: src/modules/mstr.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/mstr.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This is a single-stream to multi-stream conversion module.  What it does is accept a
 *  multi-stream M_PROTO/M_PCPROTO interface and converts it to a M_CTL/M_PCCTL interface.  It does
 *  this simply by converting M_PROTO messages to M_CTL and M_PCPROTO messages to M_PCCTL and visa
 *  versa.  The module has no service procedure on either side, but merely converts messages on the
 *  fly.  This is a very simplistic module.  M_PROTO/M_PCPROTO on the write queue are converted to
 *  M_CTL/M_PCCTL and M_CTL/M_PCCTL on the read queue are converted to M_PROTO/M_PCPROTO.  This
 *  allows user-land interface and testing of multi-stream modules that use the inter-module
 *  M_CTL/M_PCCTL interface.
 */

#include <sys/os7/compat.h>

#define MSTR_DESCRIP	"Multi-Stream (MSTR) STREAMS Module"
#define MSTR_EXTRA	"Part of Linux Fast-STREAMS"
#define MSTR_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define MSTR_REVISION	"OpenSS7 src/modules/mstr.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define MSTR_DEVICE	"SVR 4.2 MP STREAMS Multi-Stream Module"
#define MSTR_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define MSTR_LICENSE	"GPL"
#define MSTR_BANNER	MSTR_DESCRIP	"\n" \
			MSTR_EXTRA	"\n" \
			MSTR_COPYRIGHT	"\n" \
			MSTR_REVISION	"\n" \
			MSTR_DEVICE	"\n" \
			MSTR_CONTACT
#define MSTR_SPLASH	MSTR_DEVICE	" - " \
			MSTR_REVISION

#ifdef CONFIG_STREAMS_MSTR_MODULE
MODULE_AUTHOR(MSTR_CONTACT);
MODULE_DESCRIPTION(MSTR_DESCRIP);
MODULE_SUPPORTED_DEVICE(MSTR_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(MSTR_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_MSTR_MODULE */

#ifndef CONFIG_STREAMS_MSTR_NAME
#error "CONFIG_STREAMS_MSTR_NAME must be defined."
#endif				/* CONFIG_STREAMS_MSTR_NAME */
#ifndef CONFIG_STREAMS_MSTR_MODID
#error "CONFIG_STREAMS_MSTR_MODID must be defined."
#endif				/* CONFIG_STREAMS_MSTR_MODID */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-mstr");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

/*
 *  STREAMS DEFINITIONS
 */

#define MOD_ID		CONFIG_STREAMS_MSTR_MODID
#define MOD_NAME	CONFIG_STREAMS_MSTR_NAME
#ifdef MODULE
#define MOD_BANNER	MSTR_BANNER
#else				/* MODULE */
#define MOD_BANNER	MSTR_SPLASH
#endif				/* MODULE */

static struct module_info mstr_minfo = {
	.mi_idnum = MOD_ID,	/* Module ID number */
	.mi_idname = MOD_NAME,	/* Module name */
	.mi_minpsz = 0,		/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,	/* Max packet size accepted */
	.mi_hiwat = 0,		/* Hi water mark */
	.mi_lowat = 0,		/* Lo water mark */
};

static struct module_stat mstr_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat mstr_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static streamscall int mstr_qopen(queue_t *, dev_t *, int, int, cred_t *);
static streamscall int mstr_qclose(queue_t *, int, cred_t *);

static streamscall int mstr_rput(queue_t *, mblk_t *);
static streamscall int mstr_wput(queue_t *, mblk_t *);

static struct qinit mstr_rinit = {
	.qi_putp = mstr_rput,
	.qi_qopen = mstr_qopen,
	.qi_qclose = mstr_qclose,
	.qi_minfo = &mstr_minfo,
	.qi_mstat = &mstr_rstat,
};

static struct qinit mstr_winit = {
	.qi_putp = mstr_wput,
	.qi_minfo = &mstr_minfo,
	.qi_mstat = &mstr_wstat,
};

MODULE_STATIC struct streamtab mstrinfo = {
	.st_rdinit = &mstr_rinit,
	.st_wrinit = &mstr_winit,
};

/*
 *  PUT PROCEDURES
 */

STATIC streamscall int
mstr_wput(queue_t *q, mblk_t *mp)
{
	switch (__builtin_expect(mp->b_datap->db_type, M_PROTO)) {
	case M_PROTO:
		mp->b_datap->db_type = M_CTL;
		break;
	case M_PCPROTO:
		mp->b_datap->db_type = M_PCCTL;
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

STATIC streamscall int
mstr_rput(queue_t *q, mblk_t *mp)
{
	switch (__builtin_expect(mp->b_datap->db_type, M_CTL)) {
	case M_CTL:
		mp->b_datap->db_type = M_PROTO;
		break;
	case M_PCCTL:
		mp->b_datap->db_type = M_PCPROTO;
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

/*
 *  OPEN AND CLOSE
 */

STATIC streamscall int
mstr_qopen(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *crp)
{
	if (q->q_ptr != NULL)
		return (0);	/* already open */
	if (sflag != MODOPEN || WR(q)->q_next == NULL)
		return (ENXIO);
	/* dummy private structure */
	q->q_ptr = WR(q)->q_ptr = (void *) 1;
	return (0);
}

STATIC streamscall int
mstr_qclose(queue_t *q, int oflag, cred_t *crp)
{
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = NULL;
	return (0);
}

/*
 *  REGISTRATION AND INITIALIZATION
 */

#ifdef LINUX
/*
 *  Linux Registration
 */

unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, short, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for MSTR module. (0 for allocation.)");

static struct fmodsw mstr_fmod = {
	.f_name = MOD_NAME,
	.f_str = &mstrinfo,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

MODULE_STATIC int __init
mstrinit(void)
{
	int err;

	cmn_err(CE_NOTE, MOD_BANNER);
	if ((err = register_strmod(&mstr_fmod)) < 0) {
		cmn_err(CE_WARN, "%s: could not register module, err = %d", MOD_NAME, err);
		return (err);
	}
	if (modid == 0)
		modid = err;
	return (0);
}

MODULE_STATIC void __exit
mstrexit(void)
{
	int err;

	if ((err = unregister_strmod(&mstr_fmod)) < 0)
		cmn_err(CE_WARN, "%s: could not unregister module, err = %d", MOD_NAME, err);
	return;
}

module_init(mstrinit);
module_exit(mstrexit);

#endif				/* LINUX */
